File: otel-android-mobile/src/main/java/io/opentelemetry/android/mobile/export/ExportStatus.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.mobile.export

/**
 * Status of a telemetry export attempt.
 *
 * Register a listener via [ExportStatusListener] to receive these updates.
 * Useful for debugging auth failures, network issues, and buffer overflow.
 */
sealed class ExportStatus {
    /** Export succeeded. [eventCount] events were delivered. */
    data class Success(val eventCount: Int) : ExportStatus()

    /** Export failed after all retries. Events remain in buffer. */
    data class Failed(
        val reason: String,
        val eventCount: Int,
        val attempt: Int
    ) : ExportStatus()

    /** Authentication error — token invalid, expired, or blocked. */
    data class AuthError(
        val reason: String,
        val eventCount: Int
    ) : ExportStatus()

    /** Export is being retried. */
    data class Retrying(
        val attempt: Int,
        val maxAttempts: Int,
        val delayMs: Long
    ) : ExportStatus()
}

/**
 * Listener for export status events.
 *
 * Called on a background thread — do not block or perform heavy UI work.
 *
 * Usage:
 * ```kotlin
 * ExportStatusManager.addListener { status ->
 *     when (status) {
 *         is ExportStatus.AuthError -> Log.e("MyApp", "Auth failed: ${status.reason}")
 *         is ExportStatus.Failed -> analytics.track("export_failed", status.reason)
 *         is ExportStatus.Success -> {} // typically silent
 *         is ExportStatus.Retrying -> {} // optional: show progress
 *     }
 * }
 * ```
 */
fun interface ExportStatusListener {
    fun onExportStatus(status: ExportStatus)
}

/**
 * Global registry for export status listeners.
 *
 * Thread-safe. Listeners are called synchronously on the exporter thread.
 */
object ExportStatusManager {
    private val listeners = mutableListOf<ExportStatusListener>()

    fun addListener(listener: ExportStatusListener) {
        synchronized(listeners) { listeners.add(listener) }
    }

    fun removeListener(listener: ExportStatusListener) {
        synchronized(listeners) { listeners.remove(listener) }
    }

    fun clearListeners() {
        synchronized(listeners) { listeners.clear() }
    }

    internal fun notify(status: ExportStatus) {
        val snapshot: List<ExportStatusListener>
        synchronized(listeners) { snapshot = listeners.toList() }
        for (listener in snapshot) {
            try {
                listener.onExportStatus(status)
            } catch (_: Exception) {
                // Don't let listener failures break the export pipeline
            }
        }
    }
}

File: otel-ios-mobile/Sources/OTelMobileCore/Export/ExportStatus.swift
/*
 * Copyright 2025 Barry Solomon
 * SPDX-License-Identifier: Apache-2.0
 */
import Foundation

/// Observability surface for the SDK's own export pipeline. Consumers
/// (debug widget overlays, dashboards, CI smoke checks) register a
/// listener and react to each transition without having to subclass the
/// exporter. Mirrors Android's `ExportStatus` sealed class line-for-line
/// so a customer who wires a single listener implementation against both
/// platforms gets the same event stream.
///
/// `eventCount` is the number of records that were in the failing batch —
/// useful for "events lost" counters. `attempt` is 1-indexed; `delayMs`
/// is the *upcoming* sleep before the next attempt.
public enum ExportStatus: Sendable, Equatable {
    case success(eventCount: Int)
    case failed(reason: String, eventCount: Int, attempt: Int)
    /// Surfaced separately from `failed` because callers usually want to
    /// react differently — auth failure means no amount of retrying will
    /// help, the user must rotate the token.
    case authError(reason: String, eventCount: Int)
    case retrying(attempt: Int, maxAttempts: Int, delayMs: Int)
}

/// Receiver for `ExportStatus` events. Implemented as a class-bound
/// protocol so the manager can do identity-based removal (`===`).
public protocol ExportStatusListener: AnyObject {
    func onExportStatus(_ status: ExportStatus)
}

/// Process-wide registry of `ExportStatusListener`s. Pure event-driven
/// fan-out — stores no status, holds no buffer of past events. Listeners
/// register early (e.g. in the app delegate) and receive every transition
/// reported by every exporter for the rest of the process lifetime.
///
/// Thread safety: snapshot-then-iterate. The lock guards the listeners
/// array; the listener callback runs OUTSIDE the lock so a slow listener
/// can never block other notifications. Mirrors Android's pattern.
public final class ExportStatusManager: @unchecked Sendable {
    /// Process-wide instance app code uses by default. Tests construct
    /// their own instances and inject them into `RetryableExporter` so
    /// parallel suites don't cross-contaminate each other's listener
    /// transcripts.
    public static let shared = ExportStatusManager()

    private let lock = NSLock()
    private var listeners: [ExportStatusListener] = []

    public init() {}

    public func addListener(_ listener: ExportStatusListener) {
        lock.lock(); defer { lock.unlock() }
        // No dedupe — Android allows the same listener to register twice
        // and receive each event twice. Match the contract.
        listeners.append(listener)
    }

    public func removeListener(_ listener: ExportStatusListener) {
        lock.lock(); defer { lock.unlock() }
        listeners.removeAll { $0 === listener }
    }

    public func clearListeners() {
        lock.lock(); defer { lock.unlock() }
        listeners.removeAll()
    }

    /// Public so `RetryableExporter` (in the SDK target) and any user-
    /// supplied exporter can publish status. Snapshot-then-iterate so a
    /// listener that calls `removeListener` from inside its callback
    /// doesn't mutate the array we're iterating.
    public func notify(_ status: ExportStatus) {
        lock.lock()
        let snapshot = listeners
        lock.unlock()
        for listener in snapshot {
            listener.onExportStatus(status)
        }
    }
}
